CashRegister Rspec and TDD method

broimp opened the pull request:
[#115550335]

Commit:
CashRegister Rspec and TDD method
[#115550335]

Commit:
CashRegister Rspec and TDD method
[#115550335]

broimp commented:
Note, I could not get the Rspec to digest embedded double quotation marks in the method output. It escaped those characters and I couldn't figure out how to get a comparison. So I removed them from the output. If they need to be included, please explain the technique to incorporate them in the rspec test.

Commit:
Tractor class TDD development

I did not add the environment routines (.files & Gem files) since those were previously committed on the last object_tdd routine.

[#115597209]

jasonnoble commented on objects_tdd/lib/cashregister.rb:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env ruby
```
This file should be named `cash_register.rb`.

jasonnoble commented on objects_tdd/lib/cashregister.rb:
```diff
@@ -0,0 +1,34 @@
+
+# Cash Register
+class CashRegister
+  attr_accessor(:total)
```
Should the total be settable outside of the class?  This IMO should be `attr_reader :total`.

jasonnoble commented on objects_tdd/lib/cashregister.rb:
```diff
@@ -0,0 +1,34 @@
+class CashRegister
+  attr_accessor(:total)
+
+  def initialize(total = 0)
```
The initializer for CashRegister doesn't take any arguments.

jasonnoble commented on objects_tdd/lib/cashregister.rb:
```diff
@@ -0,0 +1,34 @@
+    format('$%.2f', amount)
+  end
+
+  def total
```
Get rid of this method...

jasonnoble commented on objects_tdd/lib/cashregister.rb:
```diff
@@ -0,0 +1,34 @@
+    if @total == 0
+      "pay(#{money(amount)}) # => #{money(@total)}"
+    elsif @total > 0
+      "pay(#{money(amount)}) # => Your new total is #{money(@total)}"
```
This should print "Your new total is #{money(total)}".

jasonnoble commented on objects_tdd/lib/cashregister.rb:
```diff
@@ -0,0 +1,34 @@
+    elsif @total > 0
+      "pay(#{money(amount)}) # => Your new total is #{money(@total)}"
+    else
+      "pay(#{money(amount)}) # => Your change is #{money(@total.abs)}"
```
This should print the amount of change and set the total to 0.  Add a test...

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+require 'rspec'
+require_relative '../../lib/cashregister'
```
`cash_register`

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    expect(subject).to be_a(CashRegister)
+  end
+  it 'should print a formatted total line' do
+    subject.total = 2.45
```
`subject.purchase(2.45)`

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+  it 'should print a formatted total line' do
+    subject.total = 2.45
+    test_string = 'total # => $2.45'
+    expect(subject.total).to eq(test_string)
```
`expect(subject.total).to eq(2.45)`

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+  it 'should exist' do
+    expect(subject).to be_a(CashRegister)
+  end
+  it 'should print a formatted total line' do
```
`describe '#total'`

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    test_string = 'total # => $2.45'
+    expect(subject.total).to eq(test_string)
+  end
+  it 'should add the purchase amount to the total' do
```
`describe '#purchase(amount)' do`

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    expect(subject.total).to eq(test_string)
+  end
+  it 'should add the purchase amount to the total' do
+    subject.total = 0.0
```
`expect(subject.total).to eq(0.0)`

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    subject.total = 0.0
+    subject.purchase(3.78)
+    test_string = 'total # => $3.78'
+    expect(subject.total).to eq(test_string)
```
`expect(subject.total).to eq(3.78)`

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    test_string = 'total # => $3.78'
+    expect(subject.total).to eq(test_string)
+  end
+  it 'should output a string with the purchase and the total' do
```
This wasn't one of the requirements.

jasonnoble commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    test_string = 'purchase($3.78) # => $3.78'
+    expect(subject.purchase(3.78)).to eq(test_string)
+  end
+  it 'should subtract the pay amount from the total' do
```
```
describe '#pay(amount)' do
   describe 'when paid in full' do
      it 'sets the amount to zero'
      it 'prints out "Your new balance is..."'
   describe 'when paid exact amount' do
      ...
   describe 'when overpaid' do
      ...
```

jasonnoble commented:
Needs rebasing.

Commit:
Made numerous changes per Jason’s pull request
notes. I think all issues satisfied, however,
zeroing the total on an underpayment is not
indicated since by doing so the register would
lose the ability to accept multiple payments
against an outstanding balance.

The register will zero itself only on a new
instance.
[#115550335]

Commit:
Made numerous changes per Jason’s pull request
notes. I think all issues satisfied, however,
zeroing the total on an underpayment is not
indicated since by doing so the register would
lose the ability to accept multiple payments
against an outstanding balance.

The register will zero itself only on a new
instance.
[#115550335]

wb97531 committed:
Merge pull request #15 from broimp/115597209_tractor_master

Tractor class TDD development

[#115597209]

timsjpark commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,57 @@
+    test_string = '$1.00'
+    expect(subject.total).to eq(test_string)
+  end
+  describe 'reconciliation' do
```
Generally, describe should be reserved for methods. Instead, can you use context blocks to help you organize your tests. I'm thinking something like the code below. Context blocks are very important and commonly used in the real developing world, so learning to break your tests into contexts is a valuable skill to learn.

```
describe '#pay(amount)' do
  context 'when the payment is equal to the total' do
    it 'should do some stuff' do
      # stuff
    end
  end

  context 'when the payment is less than the total' do
    it 'should do some other stuff' do
      # other stuff
    end
  end

  context 'when the payment is greater than the total' do
    it 'should do this last stuff' do
      # last stuff
    end
  end
end
```

Commit:
Added context bracketing to spec tests
[#115550335]

Commit:
Added context bracketing to spec tests
[#115550335]

Commit:
Merge branch '115550335_TDD_cash_reg_master' of github.com:broimp/learning_by_doing into 115550335_TDD_cash_reg_master
Because it is.
[#115550335]

timsjpark commented on objects_tdd/lib/cash_register.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  def total
+    money(@total.abs)
```
Is it necessary to create a `money` method? What's wrong with having the format method do the work here?

```
def total
  format('$%.2f', @total.abs)
end
```

timsjpark commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,69 @@
+  it 'should print a formatted total line' do
+    subject = CashRegister.new
+    test_string = '$0.00'
+    expect(subject.total).to eq(test_string)
```
Your test looks great! But I would argue that you don't need to keep assigning a `test_string` for each of your tests. With small strings, you can just add them into your expect statement and it will be easier to read instead of having to look at the test_string each time you want to know what you're testing for.

```
expect(subject.total).to eq('$0.00')
```

VS.

```
expect(subject.total).to eq(test_string)
```

It's just an extra line of code that, multiplied by all of your tests, adds up to half a dozen extra lines.

timsjpark commented on objects_tdd/spec/lib/cashregister_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    it 'should add the purchase amount to the total' do
+      subject = CashRegister.new
+      subject.purchase(3.78)
+      test_string = '$3.78'
```
Please refactor all of your test_strings

Commit:
Refactor spec tests, refactor object
[#115550335]

Commit:
re-apply recent changes
[#115550335]
